designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,311 @@
+
+For example, with the rule [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports), one might want to test that the rule schema fails validation when passed:
+
+- Something that isn't an array
+- An empty array
+- An array containing an item that isn't a string nor object
+- An array containing an object that is missing required properties like `name`, or has unknown properties
+- Any other invalid combinations of input
+
+Note that the goal is not to test that [ajv](https://github.com/ajv-validator/ajv) (which performs the validation using JSON Schema) itself works properly, as we can treat that third-party dependency as a blackbox, but instead to test that the rule author has written out their schema correctly.
+
+It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when no automated testing exists around this today. Schemas are often too lenient, allowing extra input that the rule implementation ignores or doesn't handle properly. This can result in rules silently misbehaving, not respecting the wishes or intentions of the user's configuration, or just crashing.
+
+By enabling testing of schemas, we'll make it easier to write higher-quality schemas and rules, thus improving the user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+Add support to the rule tester for a new `fatal` test case array in addition to `valid` and `invalid`.
+
+Example with a schema validation error because a non-existent option is passed:
+
+```json
+{
+    "fatal": [
+        {
+            "code": "foo();",
+            "options": [{ "checkFoo": true, "nonExistentOption": true }],
+            "message": "Value {\"checkFoo\":true,\"nonExistentOption\":true} should NOT have additional properties."
+        }
+    ]
+}
+```
+
+To test a custom exception thrown by a rule, simply use the exception text as the message: `"message": "Some custom exception message from the rule."`.
+
+A convenience feature to go along with this is to allow an alternative property `source: 'SCHEMA_VALIDATION' | 'RULE_IMPLEMENTATION'` to be specified indicating the type of exception without indicating the exception text.
+This would be useful since most of the time, the user may not care about the exact text of the exception, but only that the particular type of exception was thrown.
+This also makes it easier to update the exception text without having to update the messages in all the test cases.
+Alternate names for this property to consider: `messageType`, `errorType`.
```

Comment:
I think `errorType` makes sense for this. We don't really use UPPER_CASE names elsewhere, though, so I'd suggest `"schema"` or `"rule"` as the values.

Reply:
I've updated to `schema` and `rule` as the values, and `errorType` as the property name. 

My only concern is that I'm hoping it's clear that `message` and `errorType` are alternative properties that can be provided, just like `message` and `messageId`. `messageErrorType` might help group this new alternative to `message` better.

Reply:
I've also commented to suggest another approach to this property here: https://github.com/eslint/rfcs/pull/103#discussion_r1083035143

Reply:
I've updated the RFC to use an `error` object in `fatal` test cases. This object would have the `message` and/or `name` directly from the exception object. It groups the properties together nicely and enables the user to distinguish between different types of exceptions, including custom ones.

```json
{
  "error": {
    "name": "SchemaValidationError",
    "message": "...should NOT have additional properties."
  }
}
```

I believe this is significantly more elegant than the previous approach which relied on an arbitrary type property.

Reply:
I like it. 👍